### DISCUSSION TITLE
Checking whether 2 images are identical quickly

### DISCUSSION THREAD

**Discussion by Jsfasdf250** [ImageMagick/ImageMagick#6403] [Help]
### ImageMagick version

7.1.1-6 Q16-HDRI

### Operating system, version and so on

Windows 11

### Description

I would like to know if there is a way to check whether 2 images are identical or not without having to calculate every difference. If even a single difference is found, the algorithm should stop immediately.

**Comment by snibgo** [comment]
The C function `IsImagesEqual()` will stop comparing pixels as soon as any difference is found.

The CPP method ...
```
bool Magick::Image::compare(const Image &reference_) const
```
... calls `IsImagesEqual()`, so can also stop early.

At the command line, a `-subimage-search` can stop early if the metric exceeds a given threshold.
```
D:\temp>magick compare -metric RMSE -subimage-search -dissimilarity-threshold 0 imgA.miff imgB.miff NULL:
compare: images too dissimilar `imgA.miff' @ error/compare.c/CompareImagesCommand/1169.

D:\temp>echo %ERRORLEVEL%
2
```
We are searching for a subimage, but the two images are the same size, so there is only one possible location. By setting a `-dissimilarity-threshold` of zero, the search stops early if any pixels are different. Because the search did not complete, IM doesn't know the best location of the subimage, so finishes with an error.
